subsys/usb/host/usbh_desc.h

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2025 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Descriptor matching and searching utilities
+ *
+ * This file contains utilities to browse USB descriptors returned by a device
+ * according to the USB Specification 2.0 Chapter 9.
+ */
+
+#ifndef ZEPHYR_INCLUDE_USBH_DESC_H
+#define ZEPHYR_INCLUDE_USBH_DESC_H
+
+#include <stdint.h>
+
+#include <zephyr/usb/usbh.h>
+
+/**
+ * @brief Get the next descriptor in an array of descriptors.
+ *
+ * @param[in] desc Pointer to the beginning of the descriptor array to search. May be NULL.
+ * @param[in] desc_end Pointer after the end of the descriptor array
+ *
+ * @return A pointer to the first descriptor matching
+ */
+const void *usbh_desc_get_next(const void *const desc, const void *const desc_end);
+
+/**
+ * @brief Search a descriptor matching the interace number wanted.
+ *
+ * The descriptors are going to be scanned until either an interface association
+ * @c bFirstInterface field or an interface @c bInterfaceNumber field match.
+ *
+ * The descriptors following it can be browsed using @ref usbh_desc_get_next
+ *
+ * @param[in] desc Pointer to the beginning of the descriptor array to search. May be NULL.
+ * @param[in] desc_end Pointer after the end of the descriptor array
+ * @param[in] iface The interface number to search for
+ *
+ * @return A pointer to the first descriptor matching
+ */
+const void *usbh_desc_get_by_iface(const void *const desc, const void *const desc_end,
+				   const uint8_t iface);
+
+/**
+ * @brief Get the start of a device's configuration descriptor.
+ *
+ * @param[in] udev The device holding the configuration descriptor
+ *
+ * @return A pointer to the beginning of the descriptor
+ */
+const void *usbh_desc_get_cfg(const struct usb_device *const udev);
+
+/**
+ * @brief Get the end of a device's configuration descriptor.
+ *
+ * @param[in] udev The device holding the configuration descriptor
+ *
+ * @return A pointer right after the end of the descriptor (should not be accessed)
+ */
+const void *usbh_desc_get_cfg_end(const struct usb_device *const udev);
+
+/**
+ * @brief Extract information from an interface or interface association descriptors
+ *
+ * @param[in] desc The descriptor to use
+ * @param[out] iface_code Device information filled by this function
+ * @param[out] iface_num Pointer filled with the interface number, or NULL
+ *
+ * @return 0 on success or negative error code on failure.
```

Comment:
What are positive values? Please replace all the `if (ret < 0)` with `if (ret != 0)`